### Part 1: Using data found in sales.csv perform the following tasks:
    1. Standardize the Price Column
    2. Identify and remove outliers in the dataset

In [1]:
import pandas as pd
import numpy as np

In [2]:
#read sales.csv
data1 = pd.read_csv("C:/Users/Ivy/Desktop/data scientist/Python/Week 4/sales.csv")
data1.head()

Transaction_date   Product Price Payment_Type               Name  \
0    1/2/2009 6:17  Product1  1200   Mastercard           carolina   
1    1/2/2009 4:53  Product1  1200         Visa             Betina   
2   1/2/2009 13:08  Product1  1200   Mastercard  Federica e Andrea   
3   1/3/2009 14:44  Product1  1200         Visa              Gouya   
4   1/4/2009 12:56  Product2  3600         Visa            Gerd W    

                           City     State         Country   Account_Created  \
0                      Basildon   England  United Kingdom     1/2/2009 6:00   
1  Parkville                           MO   United States     1/2/2009 4:42   
2  Astoria                             OR   United States    1/1/2009 16:21   
3                        Echuca  Victoria       Australia   9/25/2005 21:13   
4  Cahaba Heights                      AL   United States  11/15/2008 15:47   

       Last_Login   Latitude   Longitude  
0   1/2/2009 6:08  51.500000   -1.116667  
1   1/2/2009 7:49  39.195000  -94.681940  
2  1/3/2009 12:32  46.188060 -123.830000  
3  1/3/2009 14:22 -36.133333  144.750000  
4  1/4/2009 12:45  33.520560  -86.802500

In [3]:
data1.describe()

Latitude   Longitude
count  997.000000  999.000000
mean    38.984398  -41.417494
std     19.499041   67.377526
min    -41.465000 -159.485280
25%     35.782500  -88.207775
50%     42.309720  -73.733890
75%     51.050000    4.850000
max     64.837780  174.766667

In [4]:
import pandas as pd
#data type for the Price column is determined since it is not included in the numerical columns above
data1 = pd.read_csv("C:/Users/Ivy/Desktop/data scientist/Python/Week 4/sales.csv")

data1['Price'].dtype

dtype('O')

In [5]:
# An error was encountered before changing the object dtype of Price to float64 showing that One value has a comma (13,000)
# Therefore,commas in values are removed using replace() function

data1 = pd.read_csv("C:/Users/Ivy/Desktop/data scientist/Python/Week 4/sales.csv")
data1['Price'] = data1['Price'].str.replace(',', '').astype(np.float64)
data1.describe()

Price    Latitude   Longitude
count    999.000000  997.000000  999.000000
mean    1650.970971   38.984398  -41.417494
std     1210.577115   19.499041   67.377526
min       20.000000  -41.465000 -159.485280
25%     1200.000000   35.782500  -88.207775
50%     1200.000000   42.309720  -73.733890
75%     1200.000000   51.050000    4.850000
max    13000.000000   64.837780  174.766667

In [6]:
import pandas as pd
import numpy as np
data1 = pd.read_csv("C:/Users/Ivy/Desktop/data scientist/Python/Week 4/sales.csv")
data1['Price'] = data1['Price'].str.replace(',', '').astype(np.float64)
data1.insert(12, "standardized_price", 0) #insert column

#Standardized Price
data1.standardized_price = (data1.Price - np.mean(data1.Price))/(np.std(data1.Price))
print(data1[['Price','standardized_price']])

#check:
mean_std = data1['standardized_price'].mean()
stdv = data1['standardized_price'].std()
print("mean:",mean_std)
print("standard deviation:",stdv)

#store to new sales.csv
data1.to_csv("C:/Users/Ivy/Desktop/data scientist/Python/Week 4/Enopia_sales.csv",index = False)

      Price  standardized_price
0    1200.0           -0.372712
1    1200.0           -0.372712
2    1200.0           -0.372712
3    1200.0           -0.372712
4    3600.0            1.610806
..      ...                 ...
995  1200.0           -0.372712
996  3600.0            1.610806
997  7500.0            4.834024
998  1200.0           -0.372712
999  1200.0           -0.372712

[1000 rows x 2 columns]
mean: -3.737417449263665e-16
standard deviation: 1.0005008765653411


In [7]:
#test for skewness
median = data1['Price'].median()
mean = data1['Price'].mean()
print("median:",median)
print("mean:",mean)

if  mean < median:
    print ("negatively skewed")
else:
    print ("positively skewed")

median: 1200.0
mean: 1650.970970970971
positively skewed


##### Two methods can be used in determining outliers. For skewed distribution, it is better to use the IQR method. For comparison purposes, results using the two methods were presented below.

In [8]:
#First method: using standard deviation(3-Sigma Limits) to remove outliers for the Price column
#This method will be applied to both price column and the standardized price column
import numpy as np

mean = np.mean(data1.Price)
std = np.std(data1.Price)
print("mean:",mean)
print("standard deviation:",std)


ll = max(0,mean - 3*std) #lower limit
ul =mean + 3*std #upper limit
print("lower limit:",ll)
print("upper limit:",ul)

cleandata1 = data1[data1.Price >=ll]
cleandata2 = data1[data1.Price <=ul]

cleandata2.describe()

mean: 1650.970970970971
standard deviation: 1209.9710691360758
lower limit: 0
upper limit: 5280.884178379198


Price    Latitude   Longitude  standardized_price
count   981.000000  979.000000  980.000000          981.000000
mean   1534.169215   39.116890  -41.097999           -0.096533
std     834.950840   19.433659   67.592075            0.690059
min      20.000000  -41.465000 -159.485280           -1.347942
25%    1200.000000   36.013475  -87.861807           -0.372712
50%    1200.000000   42.395830  -73.722083           -0.372712
75%    1200.000000   51.083333    4.916667           -0.372712
max    3600.000000   64.837780  174.766667            1.610806

##### Note:After cleaning, mean decreases to a significant value. Outliers (count) have 999-981 = 18 values

In [9]:
#First method: using standard deviation for the standardized_price column (3-Sigma Limits)

import numpy as np

mean = np.mean(data1.standardized_price)
std = np.std(data1.standardized_price)
print("mean:",mean)
print("standard deviation:",std)

ll_std = max(0,mean - 3*std) #lower limit
ul_std =mean + 3*std #upper limit
print("lower limit:",ll_std)
print("upper limit:",ul_std)

cleandata1_std = data1[data1.standardized_price >=ll_std]
cleandata2_std = data1[data1.standardized_price <=ul_std]
cleandata2_std.describe()

mean: -3.737417449263665e-16
standard deviation: 0.9999999999999999
lower limit: 0
upper limit: 2.999999999999999


Price    Latitude   Longitude  standardized_price
count   981.000000  979.000000  980.000000          981.000000
mean   1534.169215   39.116890  -41.097999           -0.096533
std     834.950840   19.433659   67.592075            0.690059
min      20.000000  -41.465000 -159.485280           -1.347942
25%    1200.000000   36.013475  -87.861807           -0.372712
50%    1200.000000   42.395830  -73.722083           -0.372712
75%    1200.000000   51.083333    4.916667           -0.372712
max    3600.000000   64.837780  174.766667            1.610806

In [10]:
#Second method: using Inter-Quartile range
IQR = np.percentile(data1.Price,75)-np.percentile(data1.Price,25)
print(IQR)

ul2=np.percentile(data1.Price,75) + 1.5*IQR
ll2=max(0,np.percentile(data1.Price,25) - 1.5*IQR)

print(ul2)
print(ll2)

nan
nan
0


##### Note: np.percentile treats nan's as very high/infinite numbers. So the high percentiles will be in the range where you will end up with a nan. To calculate the percentile without the nan, np.nanpercentile is used

In [11]:

IQR = np.nanpercentile(data1.Price,75)-np.nanpercentile(data1.Price,25)
print(IQR)

ul2=np.nanpercentile(data1.Price,75) + 1.5*IQR
ll2=max(0,np.nanpercentile(data1.Price,25) - 1.5*IQR)

print(ul2)
print(ll2)

cleandata3=data1[data1.Price >= ll2]
cleandata4=data1[data1.Price <= ul2]
cleandata4.describe()

0.0
1200.0
1200.0


Price    Latitude   Longitude  standardized_price
count   841.000000  840.000000  840.000000          841.000000
mean   1196.991677   39.228961  -41.708266           -0.375198
std      53.976284   19.221218   67.029316            0.044610
min      20.000000  -39.591667 -159.485280           -1.347942
25%    1200.000000   36.133817  -87.803677           -0.372712
50%    1200.000000   42.039165  -73.743055           -0.372712
75%    1200.000000   51.083333    4.300000           -0.372712
max    1200.000000   64.837780  174.766667           -0.372712

##### Observation: IQR removes higher number of outliers than the standard deviation. Check that cleaned data count under the price column is 981 in the first method (both using the standardized price column and the original price data) while 841 in IQR method. 
#### *** Since we have positively skewed price distribution, it is better to use the results from the IQR method***.

### Part 2: Using the data in Handling Missing Values.csv
    1.Handle all the missing values. Provide an explanation for choosing a specific method (delete, impute or keep; if impute, why mean, median or mode?) in handling specific missing values

In [12]:
#read Handling_Missing_Values.csv
data2 = pd.read_csv("C:/Users/Ivy/Desktop/data scientist/Python/Week 4/Handling_Missing_Values.csv")
data2.describe()

ID        Age        Income   Children
count  20.00000  20.000000     17.000000  18.000000
mean   10.50000  28.150000  26245.705882   1.222222
std     5.91608   1.308877    583.052395   1.437136
min     1.00000  25.000000  25094.000000   0.000000
25%     5.75000  28.000000  26037.000000   0.000000
50%    10.50000  28.000000  26234.000000   1.000000
75%    15.25000  29.000000  26666.000000   2.000000
max    20.00000  30.000000  26969.000000   4.000000

In [13]:
data2 = pd.read_csv("C:/Users/Ivy/Desktop/data scientist/Python/Week 4/Handling_Missing_Values.csv")
print(data2)

print("---------------------------------------------------------")
print("missing values per column:")
data2.isnull().sum() #determining missing values

    ID     Sex  Age   Income Employed  Children Buy_Car
0    1    Male   25  25146.0   Single       0.0      No
1    2    Male   30  26939.0  Married       2.0     Yes
2    3    Male   27  26693.0  Married       0.0      No
3    4    Male   28  26666.0  Married       3.0     Yes
4    5    Male   29  25899.0  Married       0.0      No
5    6    Male   28  26462.0  Married       1.0      No
6    7  Female   28      NaN  Married       3.0     Yes
7    8     NaN   30  26037.0  Married       2.0     Yes
8    9  Female   28  26167.0  Married       1.0     Yes
9   10     NaN   28      NaN   Single       NaN      No
10  11  Female   28  26905.0   Single       0.0      No
11  12    Male   30  26195.0   Single       NaN     Yes
12  13  Female   30  26514.0  Married       4.0     Yes
13  14  Female   28  26969.0  Married       4.0     Yes
14  15     NaN   28  26234.0  Married       0.0      No
15  16    Male   27  26229.0  Married       0.0      No
16  17  Female   26  25094.0  Married       0.0 

ID          0
Sex         4
Age         0
Income      3
Employed    0
Children    2
Buy_Car     0
dtype: int64

In [14]:
drop = data2.dropna()
print(drop)

print("---------------------------------------------------------")
print("missing values per column after dropping:")
drop.isnull().sum() #determining missing values after dropping 6 entries

    ID     Sex  Age   Income Employed  Children Buy_Car
0    1    Male   25  25146.0   Single       0.0      No
1    2    Male   30  26939.0  Married       2.0     Yes
2    3    Male   27  26693.0  Married       0.0      No
3    4    Male   28  26666.0  Married       3.0     Yes
4    5    Male   29  25899.0  Married       0.0      No
5    6    Male   28  26462.0  Married       1.0      No
8    9  Female   28  26167.0  Married       1.0     Yes
10  11  Female   28  26905.0   Single       0.0      No
12  13  Female   30  26514.0  Married       4.0     Yes
13  14  Female   28  26969.0  Married       4.0     Yes
15  16    Male   27  26229.0  Married       0.0      No
16  17  Female   26  25094.0  Married       0.0      No
18  19    Male   28  26601.0  Married       1.0      No
19  20    Male   29  25427.0  Married       1.0      No
---------------------------------------------------------
missing values per column after dropping:


ID          0
Sex         0
Age         0
Income      0
Employed    0
Children    0
Buy_Car     0
dtype: int64

##### Note: Since dropped entries consist of 30% of 20 entries, it is not feasible to do this method as well as deleting and removing data. Rather, preserving these data is a better option. For the categorical columns such as Sex and Children, missing data should be filled with "No Data" since these are personal information. It is more logical to store personal information as part of data protection for clients. However, the missing values in the income column may be replaced with estimates for statistical analysis since we only have 20 entries.

In [15]:

#preserving data
data2[['Sex','Children']]= data2[['Sex','Children']].fillna("No Data")#fill NaN vlues
print(data2)


print("---------------------------------------------------------")
print("missing values per column after filling with 'No Data':")
data2.isnull().sum()


    ID      Sex  Age   Income Employed Children Buy_Car
0    1     Male   25  25146.0   Single      0.0      No
1    2     Male   30  26939.0  Married      2.0     Yes
2    3     Male   27  26693.0  Married      0.0      No
3    4     Male   28  26666.0  Married      3.0     Yes
4    5     Male   29  25899.0  Married      0.0      No
5    6     Male   28  26462.0  Married      1.0      No
6    7   Female   28      NaN  Married      3.0     Yes
7    8  No Data   30  26037.0  Married      2.0     Yes
8    9   Female   28  26167.0  Married      1.0     Yes
9   10  No Data   28      NaN   Single  No Data      No
10  11   Female   28  26905.0   Single      0.0      No
11  12     Male   30  26195.0   Single  No Data     Yes
12  13   Female   30  26514.0  Married      4.0     Yes
13  14   Female   28  26969.0  Married      4.0     Yes
14  15  No Data   28  26234.0  Married      0.0      No
15  16     Male   27  26229.0  Married      0.0      No
16  17   Female   26  25094.0  Married      0.0 

ID          0
Sex         0
Age         0
Income      3
Employed    0
Children    0
Buy_Car     0
dtype: int64

In [16]:
#test for skewness
mn = data2['Income'].mean()
md = data2['Income'].median()
print("median:",md)
print("mean:",mn)

if  mn < md:
    print ("negatively skewed")
else:
    print ("positively skewed")

median: 26234.0
mean: 26245.70588235294
positively skewed


##### Since the income distribution is positively skewed, imputation using median was preferred.

In [17]:
median_value = data2['Income'].median() #get the median
data2['Income']=data2['Income'].fillna(median_value) #fill NaN values

#store to Cleaned_data.csv
data2.to_csv("C:/Users/Ivy/Desktop/data scientist/Python/Week 4/part2_cleaned_data.csv",index = False)

data2

ID      Sex  Age   Income Employed Children Buy_Car
0    1     Male   25  25146.0   Single      0.0      No
1    2     Male   30  26939.0  Married      2.0     Yes
2    3     Male   27  26693.0  Married      0.0      No
3    4     Male   28  26666.0  Married      3.0     Yes
4    5     Male   29  25899.0  Married      0.0      No
5    6     Male   28  26462.0  Married      1.0      No
6    7   Female   28  26234.0  Married      3.0     Yes
7    8  No Data   30  26037.0  Married      2.0     Yes
8    9   Female   28  26167.0  Married      1.0     Yes
9   10  No Data   28  26234.0   Single  No Data      No
10  11   Female   28  26905.0   Single      0.0      No
11  12     Male   30  26195.0   Single  No Data     Yes
12  13   Female   30  26514.0  Married      4.0     Yes
13  14   Female   28  26969.0  Married      4.0     Yes
14  15  No Data   28  26234.0  Married      0.0      No
15  16     Male   27  26229.0  Married      0.0      No
16  17   Female   26  25094.0  Married      0.0      No
17  18  No Data   28  26234.0   Single      0.0     Yes
18  19     Male   28  26601.0  Married      1.0      No
19  20     Male   29  25427.0  Married      1.0      No

### Part 3: Application
    1. Gather Data From Any Source (it would be better if you use your own data)
    2. Preprocess the data you gathered

In [18]:
# Read pokemon.csv (data obtained from Kaggle)
data3 = pd.read_csv("C:/Users/Ivy/Desktop/data scientist/Python/Week 4/pokemon.csv")
data3

#            Name   Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
0      1       Bulbasaur    Grass  Poison  45      49       49       65   
1      2         Ivysaur    Grass  Poison  60      62       63       80   
2      3        Venusaur    Grass  Poison  80      82       83      100   
3      4   Mega Venusaur    Grass  Poison  80     100      123      122   
4      5      Charmander     Fire     NaN  39      52       43       60   
..   ...             ...      ...     ...  ..     ...      ...      ...   
795  796         Diancie     Rock   Fairy  50     100      150      100   
796  797    Mega Diancie     Rock   Fairy  50     160      110      160   
797  798  Hoopa Confined  Psychic   Ghost  80     110       60      150   
798  799   Hoopa Unbound  Psychic    Dark  80     160       60      170   
799  800       Volcanion     Fire   Water  80     110      120      130   

     Sp. Def  Speed  Generation  Legendary  
0         65     45           1      False  
1         80     60           1      False  
2        100     80           1      False  
3        120     80           1      False  
4         50     65           1      False  
..       ...    ...         ...        ...  
795      150     50           6       True  
796      110    110           6       True  
797      130     70           6       True  
798      130     80           6       True  
799       90     70           6       True  

[800 rows x 12 columns]

In [19]:
data3 = pd.read_csv("C:/Users/Ivy/Desktop/data scientist/Python/Week 4/pokemon.csv")
data3.describe(include="all")
print(data3)

print("---------------------------------------------------------")
print("missing values per column:")
data3.isnull().sum() #determining missing values

       #            Name   Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
0      1       Bulbasaur    Grass  Poison  45      49       49       65   
1      2         Ivysaur    Grass  Poison  60      62       63       80   
2      3        Venusaur    Grass  Poison  80      82       83      100   
3      4   Mega Venusaur    Grass  Poison  80     100      123      122   
4      5      Charmander     Fire     NaN  39      52       43       60   
..   ...             ...      ...     ...  ..     ...      ...      ...   
795  796         Diancie     Rock   Fairy  50     100      150      100   
796  797    Mega Diancie     Rock   Fairy  50     160      110      160   
797  798  Hoopa Confined  Psychic   Ghost  80     110       60      150   
798  799   Hoopa Unbound  Psychic    Dark  80     160       60      170   
799  800       Volcanion     Fire   Water  80     110      120      130   

     Sp. Def  Speed  Generation  Legendary  
0         65     45           1      False  
1        

#               0
Name            1
Type 1          0
Type 2        386
HP              0
Attack          0
Defense         0
Sp. Atk         0
Sp. Def         0
Speed           0
Generation      0
Legendary       0
dtype: int64

##### From the results above,only 2 columns have missing values: Name and Type 2.Out of 800 entries, 386 values were missing in Type 2 table. It is not feasible to remove these data, therefore, missing values were replaced with "No data". Also, Type 2 consists of pokemon abilities which most are unique to every pokemon. Type 2 can be further categorized into ability types including No Data category (involving 386 pokemons).

In [20]:
data3

#            Name   Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
0      1       Bulbasaur    Grass  Poison  45      49       49       65   
1      2         Ivysaur    Grass  Poison  60      62       63       80   
2      3        Venusaur    Grass  Poison  80      82       83      100   
3      4   Mega Venusaur    Grass  Poison  80     100      123      122   
4      5      Charmander     Fire     NaN  39      52       43       60   
..   ...             ...      ...     ...  ..     ...      ...      ...   
795  796         Diancie     Rock   Fairy  50     100      150      100   
796  797    Mega Diancie     Rock   Fairy  50     160      110      160   
797  798  Hoopa Confined  Psychic   Ghost  80     110       60      150   
798  799   Hoopa Unbound  Psychic    Dark  80     160       60      170   
799  800       Volcanion     Fire   Water  80     110      120      130   

     Sp. Def  Speed  Generation  Legendary  
0         65     45           1      False  
1         80     60           1      False  
2        100     80           1      False  
3        120     80           1      False  
4         50     65           1      False  
..       ...    ...         ...        ...  
795      150     50           6       True  
796      110    110           6       True  
797      130     70           6       True  
798      130     80           6       True  
799       90     70           6       True  

[800 rows x 12 columns]

In [21]:
#preserving data
data3[['Name','Type 2']]= data3[['Name','Type 2']].fillna("No Data")#fill NaN values
print(data3)

#store to part3.csv
data3.to_csv("C:/Users/Ivy/Desktop/data scientist/Python/Week 4/Enopia_part3.csv",index = False)

print("---------------------------------------------------------")
print("missing values per column after filling with 'No Data':")
data3.isnull().sum()  #determining missing values

       #            Name   Type 1   Type 2  HP  Attack  Defense  Sp. Atk  \
0      1       Bulbasaur    Grass   Poison  45      49       49       65   
1      2         Ivysaur    Grass   Poison  60      62       63       80   
2      3        Venusaur    Grass   Poison  80      82       83      100   
3      4   Mega Venusaur    Grass   Poison  80     100      123      122   
4      5      Charmander     Fire  No Data  39      52       43       60   
..   ...             ...      ...      ...  ..     ...      ...      ...   
795  796         Diancie     Rock    Fairy  50     100      150      100   
796  797    Mega Diancie     Rock    Fairy  50     160      110      160   
797  798  Hoopa Confined  Psychic    Ghost  80     110       60      150   
798  799   Hoopa Unbound  Psychic     Dark  80     160       60      170   
799  800       Volcanion     Fire    Water  80     110      120      130   

     Sp. Def  Speed  Generation  Legendary  
0         65     45           1      False

#             0
Name          0
Type 1        0
Type 2        0
HP            0
Attack        0
Defense       0
Sp. Atk       0
Sp. Def       0
Speed         0
Generation    0
Legendary     0
dtype: int64

### -------------------END-------------------------------